File: AmicableNumbers/AN.py
#!/usr/bin/env python3

"""
Project Euler # 21

Brute Force
"""

import math

def factors(n):
    if n == 1:
        return [1]

    x = [1]
    for i in range(2, int(math.sqrt(n))+1):
        if n % i == 0:
            if i*i == n:
                x.append(i)
            else:
                x.append(i)
                x.append(n // i)
    return sorted(x)

def is_amicable(a):
    b = sum(factors(a))
    return a != b and a == sum(factors(b))


print(sum(i for i in range(10001) if is_amicable(i)))

File: AmicableNumbers/AN.java
/*
Project Euler # 21

*/

public class AN {
    public static void main( String[] args ) {
        int sum = 0;

        for ( int i=1; i<10000; i++ )
            if ( is_amicable(i) )
                sum += i;

        System.out.println(sum);
    }

    public static int factors_sum( int n ) {
        if ( n == 1 )
            return 1;

        int sum = 1;
        for ( int i=2; i<Math.round(Math.sqrt(n))+1; i++ )
            if ( n % i == 0 )
            {
                if ( i*i == n )
                    sum += i;
                else
                    sum += i + (n/i);
            }

        return sum;
    }

    public static boolean is_amicable( int a ) {
        int b = factors_sum(a);
        return  a != b && a == factors_sum(b);
    }

}
